apiserver/facades/client/application/deployrepository.go

```diff
@@ -154,45 +149,52 @@ func (api *DeployFromRepositoryAPI) DeployFromRepository(ctx context.Context, ar
 		Name:              dt.applicationName,
 		NumUnits:          dt.numUnits,
 		Placement:         dt.placement,
-		Resources:         pendingIDs,
 		Storage:           stateStorageDirectives(dt.storage),
 	}, api.store)
+	if err != nil {
+		return params.DeployFromRepositoryInfo{}, nil, []error{errors.Trace(err)}
+	}
 
-	if addApplicationErr == nil {
-		unitArgs := make([]applicationservice.AddUnitArg, dt.numUnits)
-		for i := 0; i < dt.numUnits; i++ {
-			unitName, err := unit.NewNameFromParts(dt.applicationName, nextUnitNum+i)
-			if err != nil {
-				return params.DeployFromRepositoryInfo{}, nil, []error{errors.Trace(err)}
-			}
-			unitArgs[i].UnitName = unitName
+	unitArgs := make([]applicationservice.AddUnitArg, dt.numUnits)
+	for i := 0; i < dt.numUnits; i++ {
+		unitName, err := unit.NewNameFromParts(dt.applicationName, nextUnitNum+i)
+		if err != nil {
+			return params.DeployFromRepositoryInfo{}, nil, []error{errors.Trace(err)}
 		}
-		_, addApplicationErr = api.applicationService.CreateApplication(ctx, dt.applicationName, ch, dt.origin, applicationservice.AddApplicationArgs{
-			ReferenceName: dt.charmURL.Name,
-			Storage:       dt.storage,
-			// We always have download info for a charm from the charmhub store.
-			DownloadInfo: &applicationcharm.DownloadInfo{
-				Provenance:         applicationcharm.ProvenanceDownload,
-				CharmhubIdentifier: dt.downloadInfo.CharmhubIdentifier,
-				DownloadURL:        dt.downloadInfo.DownloadURL,
-				DownloadSize:       dt.downloadInfo.DownloadSize,
-			},
-		}, unitArgs...)
+		unitArgs[i].UnitName = unitName
 	}
 
-	if addApplicationErr != nil {
-		// Check the pending resources that are added before the AddApplication is called
-		if len(pendingIDs) != 0 {
-			// Remove if there's any pending resources before raising addApplicationErr
-			removeResourcesErr := api.state.RemovePendingResources(dt.applicationName, pendingIDs, api.store)
-			if removeResourcesErr != nil {
-				api.logger.Errorf("unable to remove pending resources for %q", dt.applicationName)
-			}
+	resolvedResources := make(applicationservice.ResolvedResources, 0,
+		len(dt.resolvedResources))
+	for _, res := range dt.resolvedResources {
+		var revision *int
+		if res.Revision > 0 {
+			revision = &res.Revision
 		}
-		return params.DeployFromRepositoryInfo{}, nil, []error{errors.Trace(addApplicationErr)}
+		resolvedResources = append(resolvedResources, applicationservice.ResolvedResource{
+			Name:     res.Name,
+			Origin:   res.Origin,
+			Revision: revision,
+		})
+	}
```

Comment:
todo: This should be handled inside of api.validator.ValidateArg() where the dt structure is filled in. Looks like `resolveResources()` can be rewritten. The deployTemplate structure would be updated to have `applicationservice.ResolvedResources` rather than `pendingResourceUploads` and `resources`.

You'd also want to validate that we have all of the charm's resources accounted for there.